### PR TITLE
Gridview

### PR DESCRIPTION
**Pull request by Koerty**
TLDR : Since nobody did it, I integrated the GridView from Jacobfk20's fork

I have so much things to say about it. First, this is quite a naive implementation, as I add nearly no feature, just took code from Jacobfk20's repository, adapted and intergrated it into RetroPie's version of emulation station, and fixed the major bugs.

Their is only 2 minor bugs I didn't managed to fix (they were already here with Jacobfk20's code) : 
- pos and size arguments from theme are not taken into account
- when scrolling fast the grid look a bit cluncky (this is because the boxarts of games which aren't activly displayed are not loaded yet)

I added 2 new arguments for theme to customize the gridview : padding (between games) and scroll sound.
I plan on adding many more feature and optimizations to the gridview, but first I would like to have some feed back because this is my first pull request on this project.

EDIT : as a side effect, user will now be able to switch systems while in detailed or basic view by using left and right shoulder buttons (even if it's not displayed on the "help system" bar for detailed and basic views).

### CODE REVIEW

**Comment by lilbud** [comment]
It's good to see someone take the gridview reins from jacobfk20. My Material and Modern TV theme support grid view already if you need to test themes.

**Comment by PokeEngineer** [comment]
Nice work! I was going to do this sometime this week, but you beat me to the punch.

**Comment by hex007** [comment]
Great work mate. This should be awesome

**Comment by joolswills** [comment]
Cheers from me too. Thanks.

**Comment by dudleydes** [comment]
Thank you so much for submitting this PR. I have been using jacobfk20's Gridview mod for ES since the end of last year. I have had no issues with it but it has been frustrating to miss out on new developements and themes for ES so your efforts to integrate gridview are much appreciated.

> when scrolling fast the grid look a bit cluncky (this is because the boxarts of games which aren't activly displayed are not loaded yet)

There were issues with boxart not loading when I first used the mod but I believe they were fixed by this [commit](https://github.com/jacobfk20/EmulationStation/commit/f25ebc5c0c7039016d700bbd9251b6430f617fa6). Comparing your code in an online diff checker, I do not see these changes in your repo. Could they improve the loading of boxart?

I would be happy to test your branch with the theme I am using. Before the PR is merged, would it be a case of editing the `emulationstation.sh` scriptmodule to point to your repo rather than RetroPie's?

**Comment by dudleydes** [comment]
jacobfk20 implemented this [commit](https://github.com/jacobfk20/EmulationStation/commit/066aff5702240412fd160e66bcdc19959ab1b6be) where users could replace missing boxart with their own art. Could this be implemented? And maybe the possibility for users to use their own art for folders? Thanks.

**Comment by An0nym0usCoder** [comment]
Thanks everybody, I appreciate your enthusiasm.

I feel like this project is slowly dying because of all the forks and users going for other cooler or better looking front-ends.

So my long term goal is to give a new youth to EmulationStation by meeting today UI standards (sounds that validate or invalidate user actions, good looking (maybe some animations) and ergonomic UI ...). Ofc a big part of this features must be optional if many people want emulation station to keep a retro look to go along with their "emulation feel" (idk how to call that, sorry for my bad english)

> but it has been frustrating to miss out on new developements and themes for ES

This would be the first step of my project : take cool features from all the other EmulationStation forks and gather them here as this is one, if not the most, popular fork.
2nd step : After that I will have way better global understanding of the underlying code, so I will start develop new features (I already have some cool ideas coming to my mind, but I will also work on features asked by the community).

This lead me to the biggest problem : backward compatibility. I seriously want to know what you think about this because this look like a big problem after reading some post on RetroPie's forum talking about new features. For example in this pull-request I included 2 new theme option for the gridview : change padding between games and add scrolling sound. The thing is, themes who use this options will not be compatible anymore with older versions of EmulationStation. What should we do ? Keep an "as-it" version of themes for backward compatibility (compatible with most of EmulationStation's forks) and create a new version of each themes who want to use last up-to-date features ? This may sound ridiculous for 2 new theme options but I plan on adding way more to the gridview along with custom colors for the menu, custom icons, ... (idk if it's possible for the menu and icons I haven't checked that yet but you got the idea).

So what do you think ? Is my long-term vision compatible with RetroPie's other "main" developers ideas ? Am I asking too much from EmulationStation ?

**Comment by HerbFargus** [comment]
I don't do much with ES so I don't know if what I say has any merit but I was talking with Jools last night and he had mentioned that with all the new developments it might be worth looking at a dev branch of sorts for all the new features 

Backward compatibility is nice but at what point do we decide to require people to upgrade? The other thing is there are multiple themes in varying stages of completion (one I maintain being one of them)  I don't know if I have an answer.

Emulationstation has long been neglected so it's been good to see all of the recent interest and updates.

**Comment by lilbud** [comment]
> I already have some cool ideas coming to my mind

Yes, go on

**Comment by pjft** [comment]
I think it's fair to break retro-compatibility to a reasonable extent (meaning that those themes won't work on older versions of ES). Truth be told, if a theme with gridview is loaded in an older version it won't work either way, so this is just an extra.

There are ways to control for that, by using the theme version tag to a certain extent, but I'm not a theme expert. 

The main compatibility requirements are that old themes keep working in the latest ES versions, I don't think it's expected that things work the other way around.

A question on the "as a side effect, user will now be able to switch systems while in detailed or basic view by using left and right shoulder buttons (even if it's not displayed on the "help system" bar for detailed and basic views)" -- is that for all views, or just grid view?

Either way, thanks for putting this together.

**Comment by hex007** [comment]
Why doesn't ES just ignore tags it doesn't understand? This should be the default behaviour and it would also address the backwards compatibility issue for ever.

**Comment by pjft** [comment]
I believe ES will already ignore tags under a version it doesn't know, but
once again, that's only after versions were introduced. The problem is all
builds prior to that.

@zigurana and @jrassa will certainly be able to shed more light into that,
though, and correct me if I'm wrong as it was primarily their work.

If that's not the case, though, it could certainly be a way to make that
work.

**Comment by jrassa** [comment]
@hex007 ES doesn't ignore unknown elements, because that is how Aloshi wrote it.  A few months back there was some discussion about changing it to just log a warning, but it got sidetracked because it didn't solve the immediate issue we were trying to solve.  I would support changing this, but it wouldn't provide any immediate relief since it would only help in newer versions of ES.

@An0nym0usCoder with regards to grid view, I don't think you have to worry about backward compatibility issues with the theme changes you have made since there is no "official" ES release that supports grid view.  In that sense all of grid view is new, not just the changes you made.  That being said, there may be issues with grid view in general.  We faced a similar issue when video support was added.  The question really becomes how far back do we want to support.  I believe users running RetroPie 4.2 and newer should be safe, since there was a change that will ensure that ES completely ignores the grid view when parsing the theme.  If we want to ensure compatibility with even earlier versions, theme makers can use the `<feature>` element to hide grid view from older versions of ES.  This will require an additional code change being made.

**Comment by jrassa** [comment]
@An0nym0usCoder @jacobfk20's fork documented grid theme options here [GRIDTHEME.md](https://github.com/jacobfk20/EmulationStation/blob/Gridview/GRIDTHEME.md).  This file is missing from your fork and it looks like you have changed how the grid view is themed.  Unless I am missing something it looks like you have removed many of the options that were available.

Can you explain the thought process behind this?

**Comment by An0nym0usCoder** [comment]
@jrassa Because Jacobck20's work was just a base to start of with since I didn't know how all of this work at this time. I wanted to drop this pull request before I go too far in modifications to get some users and developers feedbacks.

As I understand better how all of this work, I'm activly rewriting big parts of the code for better looking grid, more customization options for themes and more optimized code.

PS : @jrassa @dudleydes I based my work on his repository EmulationStation-RPiE not the actual EmulationStation forked from this repository, I will double check the differences tomorrow evening.

**Comment by jrassa** [comment]
@An0nym0usCoder that would explain it.  If I remember correctly, jacobck20 original posted his work to that repository not knowing there was a RetroPie fork, then moved everything over to a new repository forked from here and continued working on it.  There are definitely additions there that were not in the repository you used.

**Comment by An0nym0usCoder** [comment]
@jrassa Thanks for the explanation I will check that later this week and report the changes

**Comment by dudleydes** [comment]
@An0nym0usCoder I last built a binary that incorporated the vram leak fix from the [Release-Beta branch](https://github.com/jacobfk20/EmulationStation/tree/Release-beta). I believe this branch has the last commits jacobfk20 made in December 2016.

**Comment by An0nym0usCoder** [comment]
This will take more time than expected, there is more changes between lastest Release-Beta branch and the EmulationStation-RPiE repository from Jacobfck20 than between EmulationStation-RPiE repository and RetroPie's repository

**Comment by Koerty** [comment]
I close this pull request because it's based on outdated code and I restarted from stratch to integrate lastest gridview version from Jacobfk20. As this version is way more compex, I may need more time than expected to integrate it.

Here is current status : 
- EmulationStation is compiling without error
- EmulationStation launch and show gridview without crashing
- Boxarts aren't showing up inside the gridview (dirty hacked for now)
- Boxarts aren't loading and unloading dynamically (dirty hacked for now)
- Some minors bugs
- It look like most of the themes settings aren't applied

**Comment by pjft** [comment]
Thanks! Do let us know if you run into problems and we can try to help
navigate the code.

Best.

**Comment by Koerty** [comment]
@pjft Can we talk in private ? Do you use any particular instant messaging app to communicate between "main" emulationstation/retropie devs ?

**Comment by pjft** [comment]
I'm not aware of anything as such. The best way is really to post it here,
as different people will have different experience with different parts of
the code.

If you do have any specific questions, I believe you may tag me in parts of
the code in this PR and I'll get a notification for it - probably the best?

Is there anything in particular you'd want to discuss in more detail?

Best, have a good weekend.

**Comment by Koerty** [comment]
@pjft Mostly general questions since I'm not used to work on open source projects, like : who is reviewing my code ? do you use any code formating conventions ?

I'm also thinking about splitting the grid view into 2 : one basic grid view with nearly only the box arts so they can take full screen width, and a detailed grid view (how the grid look at the moment, check my new pull request #203 to see a screenshot of it). The main problem is inheritance : I feel like the inheritance tree for game list views is already fucked up I don't want to make things worse, for example the function :
`remove(FileData *game, bool deleteFile)` which remove the game from the list and also from the hard drive if the boolean deleteFile is set to true. This function is defined only in BasicGameListView, and GridGameListView (with a really small difference), but not in DetailedGameListView. Shouldn't this function get moved to ISimpleGameListView or IGameListView ?

Also, why is there 2 class ISimpleGameListView and IGameListView ? Since all the game list view inherits from ISimpleGameListView, why ISimpleGameListView inherit from IGameListView and not directly from IList ? Couldn't that 2 class be merged into one ?

I feel like this project need big refactoring since everyone is adding code everywhere, pull request reviews aren't enough to ensure this project is correctly structured.

**Comment by jrassa** [comment]
> who is reviewing my code

It is a very loosely organized community.  There are not set reviewers, but there are a few of us who monitor github and try to review PR when we have the time.  I would also advise you to post on the forum.

> do you use any code formating conventions

There are no official conventions.  When editing existing files, try to keep it as close to the existing code.

> I'm also thinking about splitting the grid view into 2 : one basic grid view with nearly only the box arts so they can take full screen width, and a detailed grid view

I would advise against this.  The better option would be to give themes enough control to implement either.

> The main problem is inheritance : I feel like the inheritance tree for game list views is already fucked up I don't want to make things worse, for example the function :
remove(FileData *game, bool deleteFile) remove the game from the list and also from the hard drive if the boolean deleteFile is set to true. This function is defined only in BasicGameListView, and GridGameListView (with a really small difference), but not in DetailedGameListView. Shouldn't this function get moved to ISimpleGameListView or IGameListView ?
>
>Also, why is there 2 class ISimpleGameListView and IGameListView ? Since all the game list view inherits from ISimpleGameListView, why ISimpleGameListView inherit from IGameListView and not directly from IList ? Couldn't that 2 class be merged into one ?

Unfortunately, the original creator has been long gone so we don't really have answers to these questions.  There was probably some good intentions behind the way things are structured.

ISimpleGameListView and IGameListView probably could be merged, but just to note, neither currently inherit from IList nor should they.  The gamelist views have an IList, but they shouldn't be an IList.  I think the ultimate goal would be that there was no grid view and that themes could choose between available IList implementations (text list, image grid, image list, ...).

**Comment by Koerty** [comment]
> I would advise against this. The better option would be to give themes enough control to implement either.

I don't agree, the user should be allowed to chose if he want to see games metadatas in his grid view, it shouldn't be a themer choice. The main reason can be either user preference not to show metadatas or he just want more space on his screen for the grid so it look better on his small resolution monitor.

Same go for the number of columns and rows, a theme shouldn't be allowed to force that behavior, it should be the user job to chose so it render properly on his screen.

**Comment by pjft** [comment]
That (two different grid view modes as different views) I disagree with.
While I understand the idea and use case behind it, that opens a set of
other questions that you have to have a proper answer for, namely for the
other views.

If you want to start a shared Google document with a proposal
(architecture, etc) for review I'm sure you'll get feedback on it and it
will get attention.

However, for now, the views are static and purely controlled by themes.
You're suggesting adding an extra layer of functionality to the views to
allow for user view customization. I can imagine that you'd also want to
allow the user to hide other elements, zoom level, font size and on which
side the metadata shows, for instance.

All of those are functionally valid requirements but have so far been
addressed by the different theme makers, by design. See carbon and all its
variants for instance.

If you want to implement that, it shouldn't be considered at the grid view
level, but probably be part of a wider view refactoring.

I don't object to that effort, but let's separate them. I'd recommend
implementing grid view in accordance to the other views patterns, and then
if you want to pursue such endeavors that is a separate effortv and should
be across all views. Maybe a view modifier class or something.

I'd also refer to @zigurana's issue or PR on a similar refactoring effort
for other perspective and considerations.

In what affects views and themes, I would always heed @jdrassa's and
@zigurana's perspective as they are the people who've spent the most time
understanding that architecture.

Thanks!

**Comment by Koerty** [comment]
@pjft For now I was just thinking about 2 class, a BasicGameGridView and DetailedGameGridView, with the 2nd one inheriting from the first (a bit like BasicGameListView and DetailedGameGridView do). This doesn't change much about the structure problem since all the metadata stuff is already duplicated between the grid view and the detailed view.

I can understand the motivations behind creating a completly modulable GameView, but this would be such a huge project, even if I have the motivation to do it, this is far beyond my understanding of the project and  c++ programming skill for now.

**Comment by jrassa** [comment]
@Koerty there is an effort underway to implement a more modular gamelist view.  You can read up on it [here](https://github.com/RetroPie/EmulationStation/issues/132) and [here](https://retropie.org.uk/forum/topic/11122/testers-wanted-gamelistview-theming-refactoring).  

You say that users should be able to choose between no metadata and metadata, but that is a very binary decision.  What if users want to see some of the metadata, but not all of it?  Currently themes enable users to make that decision.  They can choose a theme that shows as much or as little metadata as they desire.  They could still do that with grid view.  They can choose a theme that shows just the amount of metadata that they want.  Even with multiple grid views, users would still be at mercy of the theme makers.  We see that now where there are some themes that do not support video and their creators are in no hurry to add support for it.  A theme maker who is only interested in grid view with metadata could be so inclined to only theme the "DetailedGridView".

With regards to the theme syntax.  That is something that I would like to see change.  I would prefer to see it more inline with how textlist and the system carousel are themed now.

**Comment by Koerty** [comment]
@jrassa Ok you are right. I will stick to minor improvements of the gridview for now. I already added scroll sound and when the user launch a game it zoom on the selected game instead of the center of the screen. I will ship that in 2 different pull request as soon as my first pull request is accepted, since they rely on it.

Plus you talked about localisation on this thread, made me think Jacobfk20 already have a traduced EmulationStation, I will take a look on how he did it and if I can integrate it in this repository.

**Comment by zigurana** [comment]
Hi @Koerty!
Thanks for joining our efforts for improving ES, it seems you have big and great idea's for the future, so let's make them happen!
I've been working on some of the refactorings you have mentioned, although I've kept grid-views completely out of scope, until such a time as there would be some activity on that front (such as: right now!).
I would suggest you work on adding the new functionalities you want add, most notably the gridview, and leave the refactorings for later. At the very least, it is a good idea to separate the submission of PRs with features from those with refactorings. If you want to join the discussion, please have a look at the [issue that @jrassa mentioned](https://github.com/RetroPie/EmulationStation/issues/132), or have a look at [my dev-branch](https://github.com/zigurana/EmulationStation/tree/ThemingRefactoring) for how I intend to approach unified gamelistviews. I would prefer it if we could include gridviews at some later stage, but I would like the gridviews to be in the 'main' ES branch first.

> The main problem is inheritance : I feel like the inheritance tree for game list views is already fucked up I don't want to make things worse, for example the function :
remove(FileData *game, bool deleteFile) remove the game from the list and also from the hard drive if the boolean deleteFile is set to true. This function is defined only in BasicGameListView, and GridGameListView (with a really small difference), but not in DetailedGameListView. Shouldn't this function get moved to ISimpleGameListView or IGameListView ?

You are completely right, I will pick this up in my refactorings activity.

> Also, why is there 2 class ISimpleGameListView and IGameListView ? Since all the game list view inherits from ISimpleGameListView, why ISimpleGameListView inherit from IGameListView and not directly from IList ? Couldn't that 2 class be merged into one ?

I am glad I am not the only one to notice! Like you said, it's a mess, and pruning out / consolidating unused classes will help with regaining some much-needed clarity.

So again, welcome, I will have a look at your other PR, let's create something great!

**Comment by Koerty** [comment]
@zigurana I don't know what to answer, so I will just say #makeEmulationStationGreatAgain
